generators/chipyard/src/main/scala/IOBinders.scala

```diff
@@ -0,0 +1,167 @@
+})
+
+class WithBlockDeviceModel extends RegisterIOBinder({
+  (c, r, s, top: CanHavePeripheryBlockDeviceModuleImp) => top.connectBlockDeviceModel(); Nil
+})
+
+class WithLoopbackNIC extends RegisterIOBinder({
+  (c, r, s, top: CanHavePeripheryIceNICModuleImp) => top.connectNicLoopback(); Nil
+})
+
+class WithUARTAdapter extends RegisterIOBinder({
+  (c, r, s, top: HasPeripheryUARTModuleImp) => {
+    val defaultBaudRate = 115200 // matches sifive-blocks uart baudrate
+    top.uart.zipWithIndex.foreach{ case (dut_io, i) =>
+      val uart_sim = Module(new UARTAdapter(i, defaultBaudRate)(top.p))
+      uart_sim.io.uart.txd := dut_io.txd
+      dut_io.rxd := uart_sim.io.uart.rxd
+    }
+    Nil
+  }
+})
+
+class WithSimAXIMem extends RegisterIOBinder({
+  (c, r, s, top: CanHaveMasterAXI4MemPortModuleImp) => top.connectSimAXIMem(); Nil
+})
+
+class WithSimAXIMMIO extends RegisterIOBinder({
+  (c, r, s, top: CanHaveMasterAXI4MMIOPortModuleImp) => top.connectSimAXIMMIO(); Nil
+})
+
+class WithDontTouchPorts extends RegisterIOBinder({
+  (c, r, s, top: DontTouch) => top.dontTouchPorts(); Nil
+})
+
+class WithTieOffInterrupts extends RegisterIOBinder({
+  (c, r, s, top: HasExtInterruptsBundle) => top.tieOffInterrupts(); Nil
+})
+
+class WithTieOffL2FBusAXI extends RegisterIOBinder({
+  (c, r, s, top: CanHaveSlaveAXI4PortModuleImp) => {
+    top.l2_frontend_bus_axi4.foreach(axi => {
+      axi.tieoff()
+      experimental.DataMirror.directionOf(axi.ar.ready) match {
+        case ActualDirection.Input =>
+          axi.r.bits := DontCare
+          axi.b.bits := DontCare
+        case ActualDirection.Output =>
+          axi.aw.bits := DontCare
+          axi.ar.bits := DontCare
+          axi.w.bits := DontCare
+      }
+    })
+    Nil
+  }
+})
+
+class WithTiedOffDebug extends RegisterIOBinder({
+  (c, r, s, top: HasPeripheryDebugModuleImp) => {
+    Debug.tieoffDebug(top.debug, top.psd)
+    // tieoffDebug doesn't actually tie everything off :/
+    top.debug.foreach(_.clockeddmi.foreach({ cdmi => cdmi.dmi.req.bits := DontCare }))
+    Nil
+  }
+})
+
+class WithSimSerial extends RegisterIOBinder({
+  (c, r, s, top: CanHavePeripherySerialModuleImp) => {
+    val ser_success = top.connectSimSerial()
+    when (ser_success) { s := true.B }
+    Nil
+  }
+})
+
+class WithTiedOffSerial extends RegisterIOBinder({
+  (c, r, s, top: CanHavePeripherySerialModuleImp) => top.tieoffSerial(); Nil
+})
+
+
+class WithSimDTM extends RegisterIOBinder({
+  (c, r, s, top: HasPeripheryDebugModuleImp) => {
+    val dtm_success = Wire(Bool())
+    top.reset := r | top.debug.map { debug => AsyncResetReg(debug.ndreset) }.getOrElse(false.B)
+    Debug.connectDebug(top.debug, top.psd, c, r, dtm_success)(top.p)
+    when (dtm_success) { s := true.B }
+    Nil
+  }
+})
+
+
+class WithTraceGenSuccessBinder extends RegisterIOBinder({
+  (c, r, s, top: HasTraceGenTilesModuleImp) => s := top.success; Nil
```

Comment:
We need to be careful with last connect semantics here.
If the other success functions from `WithSimDTM` or `WithSimSerial` are before this then this direct assign (no when) will overwrite both of them.
I would make them all `when`s and also make sure there is no direct assign in `Top`.

Reply:
What do you mean no direct assign? As in wrap all assigns to `Top` IOs in `when`s?

Reply:
At least success which we know could be problematic.

Reply:
Done